src/main/java/net/snowflake/ingest/streaming/internal/ArrowRowBuffer.java

```diff
@@ -465,45 +465,47 @@ private float convertRowToArrow(
       } else {
         switch (logicalType) {
           case FIXED:
-            if (!field.getMetadata().get(COLUMN_SCALE).equals("0")
-                || physicalType == ColumnPhysicalType.SB16) {
-              int scale =
-                  DataValidationUtil.validateAndParseInteger(field.getMetadata().get(COLUMN_SCALE));
-              BigDecimal bigDecimalValue = DataValidationUtil.validateAndParseBigDecimal(value);
-
-              // vector.setSafe requires the BigDecimal input scale explicitly match its scale
-              bigDecimalValue = bigDecimalValue.setScale(scale);
-              ((DecimalVector) vector).setSafe(curRowIndex, bigDecimalValue);
-              BigInteger intRep =
-                  bigDecimalValue
-                      .multiply(BigDecimal.valueOf(Power10.intTable[scale]))
-                      .toBigInteger();
-              stats.addIntValue(intRep);
+            int columnPrecision = Integer.parseInt(field.getMetadata().get(COLUMN_PRECISION));
+            int columnScale = getColumnScale(field.getMetadata());
+            BigDecimal inputAsBigDecimal = DataValidationUtil.validateAndParseBigDecimal(value);
+            // vector.setSafe requires the BigDecimal input scale explicitly match its scale
+            inputAsBigDecimal = inputAsBigDecimal.setScale(columnScale, RoundingMode.HALF_UP);
+
+            if (inputAsBigDecimal.abs().compareTo(BigDecimal.TEN.pow(columnPrecision - columnScale))
+                >= 0) {
+              throw new SFException(
+                  ErrorCode.INVALID_ROW,
+                  inputAsBigDecimal,
+                  String.format(
+                      "Number out of representable exclusive range of (-1e%s..1e%s)",
+                      columnPrecision - columnScale, columnPrecision - columnScale));
+            }
+
+            if (columnScale != 0 || physicalType == ColumnPhysicalType.SB16) {
+              ((DecimalVector) vector).setSafe(curRowIndex, inputAsBigDecimal);
+              stats.addIntValue(inputAsBigDecimal.unscaledValue());
               rowBufferSize += 16;
             } else {
               switch (physicalType) {
                 case SB1:
-                  byte byteValue = DataValidationUtil.validateAndParseByte(value);
-                  ((TinyIntVector) vector).setSafe(curRowIndex, byteValue);
-                  stats.addIntValue(BigInteger.valueOf(byteValue));
+                  ((TinyIntVector) vector).setSafe(curRowIndex, inputAsBigDecimal.byteValueExact());
```

Comment:
will `inputAsBigDecimal.byteValueExact()` throw if `inputAsBigDecimal` is outside of byte range?
if so, do we need to wrap it into `SFException`?
cc @sfc-gh-tzhang

Reply:
I don't think so, I believe we have a big catch block in `insertRows` to catch everything and throw a SFException instead

Reply:
We have a try-catch block in insertRows, which is catching not just SFException, but all exceptions. However, this will not happen because above in the `convertRowToArrow` method, there is now validation based on logical type, so if the value is outside of the physical byte range, for example, it would be also outside of logical range and the validation would catch it.